lock-api/src/main/java/com/palantir/lock/watch/LockWatchEventCacheImpl.java

```diff
@@ -0,0 +1,273 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.lock.watch;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import org.immutables.value.Value;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Ordering;
+import com.google.common.collect.TreeMultimap;
+import com.google.errorprone.annotations.concurrent.GuardedBy;
+import com.palantir.lock.LockDescriptor;
+import com.palantir.lock.v2.LockToken;
+import com.palantir.logsafe.Preconditions;
+
+public final class LockWatchEventCacheImpl implements LockWatchEventCache {
+    private final ClientLockWatchEventLog eventLog;
+    private final HashMap<Long, MapEntry> timestampMap = new HashMap<>();
+    private final TreeMultimap<IdentifiedVersion, Long> aliveVersions =
+            TreeMultimap.create(IdentifiedVersion.comparator(), Ordering.natural());
+
+    @GuardedBy("this")
+    private boolean failed = false;
```

Comment:
Similarly to previous comments, you probably want this implemented just once and poison the entire thing.

Reply:
Done